src/workerd/api/unsafe.c++

```diff
@@ -0,0 +1,46 @@
+#include "unsafe.h"
+
+namespace workerd::api {
+
+namespace {
+static constexpr auto EVAL_STR = "eval"_kjc;
+inline kj::StringPtr getName(jsg::Optional<kj::String>& name, kj::StringPtr def) {
+  return name.map([](kj::String& str) {
+    return str.asPtr();
+  }).orDefault(def);
+}
+}  // namespace
+
+jsg::JsValue UnsafeEval::eval(jsg::Lock& js, kj::String script,
+                              jsg::Optional<kj::String> name) {
+  js.setAllowEval(true);
```

Comment:
@jasnell Is this call actually necessary? Since we're not actually invoking the JS eval API in this code -- we're calling the C++ APIs directly -- I suspect there's no need to enable eval here?

Reply:
This is specifically to allow for cases where the script being eval'd itself contains nested regular `eval()` expressions, e.g.

```
env.unsafe.eval('eval(1)');
```

While this example is a bit silly, it's not unlikely that nested evals could be included in the script passed into the `unsafe.eval(...)`

Reply:
I didn't realize that that was the case @jasnell. I think it's undesirable to recursively allow eval to be used by unprivileged code.

We use dynamic execution in a specialty/privileged code that powers local development workflows/ Outside of this specialty code, we don't want people to use eval as that could user provided code to run locally but not in production.

I propose that we remove this recursive feature. If the privileged code has a need to expose eval to userland code, it can do so via `globalThis.eval = env.unsafe.eval` before evaling user's code.

Reply:
`globalThis.eval = env.unsafe.eval` wouldn't actually work directly, that would end up with a "Illegal invocation" error, but you can get close with `globalThis.eval = function(code) { return env.unsafe.eval(code); }`

> We use dynamic execution in a specialty/privileged code that powers local development workflows/ Outside of this specialty code, we don't want people to use eval as that could user provided code to run locally but not in production.

I'm not sure how this would allow use of eval outside of this scope? The regular `eval()` would *only* be allowed *within* the script passed into `env.unsafe.eval(...)`, which is only enabled with the binding that only exists in the local runtime. There's nothing here that would allow that to be used in production.

For instance, even with the binding enabled, the following still throws an error:

```
eval('1+1');  // throws!
```

But 

```
env.unsafe.eval('eval("1+1")');  // works
```

Even if I tried something silly like the following, trying to break out of the restriction we still get an appropriate error.

```
const fn = env.unsafe.eval(`
function foo() { eval('1+1'); }
foo;
`);

fn();  // throws!
```

Reply:
I'm not sure I understand why this is a problem. Using `Ajv` like this at the top level, the user would still need to call `unsafe.eval(...)` to use it, and we would only allow for `eval(...)` and `new Function(...)` within the synchronous execution of that call. *ANY* worker code that is using `unsafe.eval(...)`, whether it contains a nested eval or not, will not work when deployed to production because we cannot deploy a worker with the `unsafe` binding at all -- it simply doesn't and won't exist outside of the local workerd environment.